### PR TITLE
[INTEL MKL] Reimplement CompareMklDnnLayouts

### PR DESCRIPTION
**Pull request by Solaryee**
The implementation of `CompareMklDnnLayouts` in mkl_util.h was not correct. The code compared `mkldnn_memory_desc_t` by bytes and didn't exclude the garbage values, which causing the comparison always failing thus unnecessary reorders were executed. We reimplement the function by comparing class members one by one.

### CODE REVIEW

File: tensorflow/core/util/mkl_util.h [hunk 1]
```diff
@@ -296,14 +337,23 @@ class MklDnnShape {
...
-    const char* d2 = reinterpret_cast<const char*>(&mdd2);
 
-    size_t md_size = sizeof(mdd1);
-    for (size_t i = 0; i < md_size; i++) {
-      if (*d1++ != *d2++) {
-        return false;
-      }
+    assert(mdd1.primitive_kind == mkldnn::primitive::kind::memory);
+    assert(mdd2.primitive_kind == mkldnn::primitive::kind::memory);
+    bool base_equal = true && mdd1.ndims == mdd2.ndims &&
```
Comment (penpornk) [inline tensorflow/core/util/mkl_util.h]: Nit: We don't need `true`.
```suggestion
    bool base_equal = mdd1.ndims == mdd2.ndims &&
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR! Just curious, why does the struct has garbage in the first place?

**Comment by ShengYang1** [comment]
> Thank you for the PR! Just curious, why does the struct has garbage in the first place?

For example, `dims` field in `mkldnn_memory_desc_t` is a C array with size 12. Most of the time, only the first four values are used and the rest are uninitialized. So we couldn't compare it by bytes.

File: tensorflow/core/util/mkl_util.h (hunk 1 above)
Comment (ShengYang1) [reply tensorflow/core/util/mkl_util.h]: Done

**Comment by ShengYang1** [comment]
Any updates?

**Review by penpornk — APPROVED** [review]
I'm so sorry for the delay! Thank you again for the PR!

**Comment by penpornk** [comment]
> > Thank you for the PR! Just curious, why does the struct has garbage in the first place?
> 
> For example, `dims` field in `mkldnn_memory_desc_t` is a C array with size 12. Most of the time, only the first four values are used and the rest are uninitialized. So we couldn't compare it by bytes.

Thanks for the explanation as well! :)

**Pull request merged**